src/main/java/com/commercetools/sync/sdk2/products/helpers/VariantReferenceResolver.java

```diff
@@ -0,0 +1,256 @@
+   * contains a new instance of the draft with the resolved references.
+   *
+   * <p>Note: this method will filter out any null sub resources (e.g. prices, attributes or assets)
+   * under the returned resolved variant.
+   *
+   * @param productVariantDraft the product variant draft to resolve it's references.
+   * @return a {@link java.util.concurrent.CompletionStage} that contains as a result a new
+   *     productDraft instance with resolved references or, in case an error occurs during reference
+   *     resolution, a {@link
+   *     com.commercetools.sync.commons.exceptions.ReferenceResolutionException}.
+   */
+  @Override
+  public CompletionStage<ProductVariantDraft> resolveReferences(
+      @Nonnull final ProductVariantDraft productVariantDraft) {
+    return resolvePricesReferences(ProductVariantDraftBuilder.of(productVariantDraft))
+        .thenCompose(this::resolveAssetsReferences)
+        .thenCompose(this::resolveAttributesReferences)
+        .thenApply(ProductVariantDraftBuilder::build);
+  }
+
+  @Nonnull
+  CompletionStage<ProductVariantDraftBuilder> resolveAssetsReferences(
+      @Nonnull final ProductVariantDraftBuilder productVariantDraftBuilder) {
+
+    final List<AssetDraft> productVariantDraftAssets = productVariantDraftBuilder.getAssets();
+    if (productVariantDraftAssets == null) {
+      return completedFuture(productVariantDraftBuilder);
+    }
+
+    return mapValuesToFutureOfCompletedValues(
+            productVariantDraftAssets, assetReferenceResolver::resolveReferences, toList())
+        .thenApply(productVariantDraftBuilder::assets);
+  }
+
+  @Nonnull
+  CompletionStage<ProductVariantDraftBuilder> resolvePricesReferences(
+      @Nonnull final ProductVariantDraftBuilder productVariantDraftBuilder) {
+
+    final List<PriceDraft> productVariantDraftPrices = productVariantDraftBuilder.getPrices();
+    if (productVariantDraftPrices == null) {
+      return completedFuture(productVariantDraftBuilder);
+    }
+
+    return mapValuesToFutureOfCompletedValues(
+            productVariantDraftPrices, priceReferenceResolver::resolveReferences, toList())
+        .thenApply(productVariantDraftBuilder::prices);
+  }
+
+  @Nonnull
+  private CompletionStage<ProductVariantDraftBuilder> resolveAttributesReferences(
+      @Nonnull final ProductVariantDraftBuilder productVariantDraftBuilder) {
+
+    final List<Attribute> attributeDrafts = productVariantDraftBuilder.getAttributes();
+    if (attributeDrafts == null) {
+      return completedFuture(productVariantDraftBuilder);
+    }
+
+    return mapValuesToFutureOfCompletedValues(
+            attributeDrafts, this::resolveAttributeReference, toList())
+        .thenApply(productVariantDraftBuilder::attributes);
+  }
+
+  @Nonnull
+  private CompletionStage<Attribute> resolveAttributeReference(
+      @Nonnull final Attribute attributeDraft) {
+    final Object value = attributeDraft.getValue();
+
+    if (!(value instanceof JsonNode)) {
```

Comment:
This will always be true as the attributeValue is never instance of JsonNode but Object.